kernel/src/engine/arrow_expression/in_list.rs

```diff
@@ -0,0 +1,572 @@
+        ArrowDataType::Int8 => eval_arrow(left_as!(Int8Type)?, &right_arr)?,
+        ArrowDataType::Int16 => eval_arrow(left_as!(Int16Type)?, &right_arr)?,
+        ArrowDataType::Int32 => eval_arrow(left_as!(Int32Type)?, &right_arr)?,
+        ArrowDataType::Int64 => eval_arrow(left_as!(Int64Type)?, &right_arr)?,
+        ArrowDataType::Float16 => eval_arrow(left_as!(Float16Type)?, &right_arr)?,
+        ArrowDataType::Float32 => eval_arrow(left_as!(Float32Type)?, &right_arr)?,
+        ArrowDataType::Float64 => eval_arrow(left_as!(Float64Type)?, &right_arr)?,
+        ArrowDataType::Timestamp(TimeUnit::Microsecond, _) => {
+            eval_arrow(left_as!(TimestampMicrosecondType)?, &right_arr)?
+        }
+        ArrowDataType::Date32 => eval_arrow(left_as!(Date32Type)?, &right_arr)?,
+        ArrowDataType::Decimal128(_, _) => eval_arrow(left_as!(Decimal128Type)?, &right_arr)?,
+        _ => {
+            return Err(Error::invalid_expression(format!(
+                "Cannot compare {} to {}",
+                left_arr.data_type(),
+                list_field.data_type()
+            )))
+        }
+    })
+}
+
+fn is_in_list(ad: &ArrayData, values: impl IntoIterator<Item = Option<Scalar>>) -> BooleanArray {
+    #[allow(deprecated)]
+    let inlist = ad.array_elements();
+    // `v IN (k1, ..., kN)` is logically equivalent to `v = k1 OR ... OR v = kN`, so evaluate
+    // it as such, ensuring correct handling of NULL inputs (including `Scalar::Null`).
+    values
+        .into_iter()
+        .map(|v| {
+            KernelPredicateEvaluatorDefaults::finish_eval_junction(
+                JunctionOperator::Or,
+                inlist
+                    .iter()
+                    .map(|k| Some(v.as_ref()?.partial_cmp(k)? == Ordering::Equal)),
+                false,
+            )
+        })
+        .collect()
+}
+
+// helper function to make arrow in_list* kernel results comliant with SQL NULL semantics.
+// Specifically, if an item is not found in the in-list, but the in-list contains NULLs, the
+// result should be NULL (UNKNOWN) as well.
+fn fix_arrow_in_list_result(
+    results: BooleanArray,
+    in_lists: impl IntoIterator<Item = Option<ArrayRef>>,
+) -> BooleanArray {
+    results
+        .iter()
+        .zip(in_lists)
+        .map(|(res, arr)| match (res, arr) {
+            (Some(false), Some(arr)) if arr.null_count() > 0 => None,
+            _ => res,
+        })
+        .collect()
+}
+
+fn eval_arrow_utf8(
+    left_arr: &dyn Array,
+    right_arr: &GenericListArray<i32>,
+) -> Result<BooleanArray, Error> {
+    let result = match left_arr.data_type() {
+        ArrowDataType::Utf8 => in_list_utf8(left_arr.as_string::<i32>(), right_arr),
+        ArrowDataType::LargeUtf8 => in_list_utf8(left_arr.as_string::<i64>(), right_arr),
+        _ => return Err(Error::invalid_expression("Expected string column")),
+    }
+    .map_err(Error::generic_err)?;
+    Ok(fix_arrow_in_list_result(result, right_arr.iter()))
```

Comment:
Isn't this redundant, when we already `impl From<ArrowError> for Error`?
```suggestion
    Ok(fix_arrow_in_list_result(result?, right_arr.iter()))
```
(lots of these)